### ISSUE TITLE
Error in TransposedConv2d

### ISSUE THREAD

**Issue by Shashi456** [tensorflow/swift-apis#282]
Refer to comment [here](https://github.com/tensorflow/swift-apis/issues/53#issuecomment-504767738).

Even when trying to write test cases for #174 and the transposed conv2d layer, I found a similar error and i reported it there. I think we need to add preconditions and check the layer logic thoroughly.

**Comment by Shashi456** [comment]
The error in question : 

> 
> Trying to use the TransposedConv2D layer but getting strange error. Below is a small example that when I run either in Colab or locally with a recent build, I get the following error. Any ideas??
> 
> [1, 4, 4, 256]
> 0.69032013
> Fatal error: Conv2DCustomBackpropFilter: filter and out_backprop must have the same out_depth: file /swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2123
> Current stack trace:
> 0    libswiftCore.so                    0x00007f06347314a0 _swift_stdlib_reportFatalErrorInFile + 115
> 1    libswiftCore.so                    0x00007f063467930c <unavailable> + 3035916
> 2    libswiftCore.so                    0x00007f06346793fe <unavailable> + 3036158
> 3    libswiftCore.so                    0x00007f06344c06c2 <unavailable> + 1230530
> 4    libswiftCore.so                    0x00007f0634646292 <unavailable> + 2826898
> 5    libswiftCore.so                    0x00007f06344bfba9 <unavailable> + 1227689
> 6    libswiftTensorFlow.so              0x00007f0631aae572 <unavailable> + 599410
> 7    libswiftTensorFlow.so              0x00007f0631aaccc0 checkOk(_:file:line:) + 508
> 8    libswiftTensorFlow.so              0x00007f0631ad1ad0 _TFCCheckOk(_:) + 81
> 9    libswiftTensorFlow.so              0x00007f0631ad1ac0 _swift_tfc_CheckOk + 9
> Current stack trace:
> 	frame #12: 0x00007f0661d90ac9 $__lldb_expr76`AD__$s14__lldb_expr_7510TensorFlow0C0VySfGAB16TransposedConv2DVXEfU___adjoint_src_0_wrt_0 at <Cell 14>:14:15
> 	frame #20: 0x00007f0661d8e9a0 $__lldb_expr76`main at <Cell 14>:13:38
> 	frame #33: 0x00007f0661d98be7 $__lldb_expr72`AD__$s14__lldb_expr_7110TensorFlow0C0VySfGAB16TransposedConv2DVXEfU___primal_src_0_wrt_0(layer=<unavailable>) at <Cell 13>:14:15
> 	frame #34: 0x00007f0661d99b84 $__lldb_expr72`AD__$s14__lldb_expr_7110TensorFlow0C0VySfGAB16TransposedConv2DVXEfU___vjp_src_0_wrt_0 at <Cell 13>:13:56
> 	frame #42: 0x00007f0661d979a0 $__lldb_expr72`main at <Cell 13>:13:38
> 	frame #55: 0x00007f0661da2be7 $__lldb_expr68`AD__$s14__lldb_expr_6710TensorFlow0C0VySfGAB16TransposedConv2DVXEfU___primal_src_0_wrt_0(layer=<unavailable>) at <Cell 12>:14:15
> 	frame #56: 0x00007f0661da3b84 $__lldb_expr68`AD__$s14__lldb_expr_6710TensorFlow0C0VySfGAB16TransposedConv2DVXEfU___vjp_src_0_wrt_0 at <Cell 12>:13:56
> 	frame #64: 0x00007f0661da19a0 $__lldb_expr68`main at <Cell 12>:13:38
> 
> import TensorFlow
> 
> let tensorShape = TensorShape([4, 4, 256, 64])
> let seed = (Int64.random(in: Int64.min..<Int64.max), Int64.random(in: Int64.min..<Int64.max))
> let transposedConv2D = TransposedConv2D(
>   filter: Tensor(glorotUniform: tensorShape, seed: seed),
>   bias: Tensor(zeros: TensorShape([128])),
>   activation: identity,
>   strides: (1, 1),
>   padding: .valid)
> 
> let tensor = Tensor<Float>(randomUniform: TensorShape([1,1,1,64]))
> let (loss, grads) = transposedConv2D.valueWithGradient { layer -> Tensor<Float> in
>     let out = layer(tensor)
>     print(out.shape)
>     let loss = out.sum()
>     print(loss)
>     return loss
> }
>

**Comment by sjaz24** [comment]
@Shashi456 do you think it has to do with this. Shouldn't channels (c) be filter.shape[3]. I'm looking at the original merged code, so not sure???

```
@differentiable
    public func applied(to input: Tensor<Float>, in _: Context) -> Tensor<Float> {
        let batchSize = input.shape[0]
        let w = (input.shape[1] - (1 * paddingIndex)) * strides.0 + (filter.shape[0] * paddingIndex)
        let h = (input.shape[2] - (1 * paddingIndex)) * strides.1 + (filter.shape[1] * paddingIndex)
        let c = filter.shape[2]
        let newShape = Tensor<Int32>([batchSize, w, h, c])
        return activation(input.conv2DBackpropInput(shape: newShape, filter: filter,
                                                         strides: (1, strides.0, strides.1, 1),
                                                         padding: padding) + bias)
    }
```

**Comment by sjaz24** [comment]
Note this other issue with this layer. https://bugs.swift.org/browse/TF-540

**Comment by Shashi456** [comment]
The solution also seems to be here, https://groups.google.com/a/tensorflow.org/forum/m/#!msg/swift/UUPwV01sZrE/LszG6T7dBQAJ


@sjaz24, would you be up to fix the error and write a test for the layer? I think it's the perfect opportunity for a first contribution to the repo :) , if not I'll put a PR up later today.

**Comment by sjaz24** [comment]
I've already done what he suggests locally to get by that issue but then I run into this issue so I believe them to be different issues but wanted to point out the issue in case it's related. I'll play around with it and see. I really find this line below weird. Hmmm.. I'll give it a go at fixing it.

let c = filter.shape[2]

**Comment by Shashi456** [comment]
@sjaz24 maybe you could take a look at the tests I wrote in #174, the filter dimensions and other dimensions were inspired by running a transposed layer in tensor flow, I ran into the filter depth and out backprop error there as well and just played around with the dimensions until they worked.  Maybe that could be another starting point.

**Comment by Shashi456** [comment]
@rxwei I had some doubts, when i'm trying to solve this, Should i open a PR and we could continue the conversation and discuss the solution there?

**Comment by eaplatanios** [comment]
@Shashi456 Feel free to open a PR and continue the conversation there.
